In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sqlalchemy import create_engine

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
housing_df = pd.read_sql_query('select * from houseprices',con=engine)


engine.dispose()

In [2]:
#Create dummies for the categorical features

housing_df = pd.concat([housing_df,pd.get_dummies(housing_df.kitchenqual, prefix='kitchenqual', drop_first=True)], axis=1)
housing_df = pd.concat([housing_df,pd.get_dummies(housing_df.exterqual, prefix='exterqual', drop_first=True)], axis=1)

dummy_column_names = list(pd.get_dummies(housing_df.kitchenqual, prefix='kitchenqual', drop_first=True).columns)
dummy_column_names = dummy_column_names + list(pd.get_dummies(housing_df.exterqual, prefix='exterqual', drop_first=True).columns)

In [3]:
#Setting variables for model
X = housing_df[['overallqual', 'grlivarea', 'garagecars'] + dummy_column_names]
Y = housing_df.saleprice

In [4]:
#Running regression model

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.779
Model:                            OLS   Adj. R-squared:                  0.777
Method:                 Least Squares   F-statistic:                     567.3
Date:                Sun, 18 Aug 2019   Prob (F-statistic):               0.00
Time:                        12:02:19   Log-Likelihood:                -17443.
No. Observations:                1460   AIC:                         3.491e+04
Df Residuals:                    1450   BIC:                         3.496e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const           5.846e+04   1.13e+04      5.189      0.000    3.64e+04    8.06e+04
overallqual     1.768e+04   1236.605     14.299      0.000    1.53e+04    2.01e+04
grlivarea         48.6266      2.369     20.530      0.000      43.980      53.273
garagecars      1.856e+04   1714.192     10.830      0.000    1.52e+04    2.19e+04
kitchenqual_Fa -5.282e+04   8556.765     -6.172      0.000   -6.96e+04    -3.6e+04
kitchenqual_Gd -4.101e+04   4909.002     -8.354      0.000   -5.06e+04   -3.14e+04
kitchenqual_TA -5.139e+04   5510.696     -9.326      0.000   -6.22e+04   -4.06e+04
exterqual_Fa   -5.861e+04   1.32e+04     -4.438      0.000   -8.45e+04   -3.27e+04
exterqual_Gd   -4.153e+04   6564.259     -6.326      0.000   -5.44e+04   -2.87e+04
exterqual_TA    -5.43e+04   7245.726     -7.494      0.000   -6.85e+04   -4.01e+04
==============================================================================
Omnibus:                      362.267   Durbin-Watson:                   2.002
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            21126.953
Skew:                           0.055   Prob(JB):                         0.00
Kurtosis:                      21.635   Cond. No.                     2.69e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.69e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC. Do you think your model is satisfactory? If so, why?
R-squared = 0.779 <br>
Adjusted R-squared = 0.777 <br>
F-test = 567.3, p-value = 0.00 <br>
AIC = 34910 <br>
BIC = 34960 <br><br>

The F-test indicates that this model has variances explained by the features, and the adjusted R-squared value of 0.777 indicates that 77.7% of the variance in the target is explained by the explanatory variables.  As there is only one model so far, the AIC and BIC scores aren't helpful because there are no models to compare this one to.  Overall, I think that this model is satisfactory, but there is still room for improvement since 22.3% of the variance in the target is unexplained. 

In [5]:
#Setting additional variables for model 2: garagearea and mszoning
housing_df = pd.concat([housing_df,pd.get_dummies(housing_df.mszoning, prefix='mszoning', drop_first=True)], axis=1)
dummy_column_names = dummy_column_names + list(pd.get_dummies(housing_df.mszoning, prefix='mszoning', drop_first=True).columns)

X = housing_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea'] + dummy_column_names]
Y = housing_df.saleprice

In [6]:
#Running regression model 2

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.792
Model:                            OLS   Adj. R-squared:                  0.790
Method:                 Least Squares   F-statistic:                     392.6
Date:                Sun, 18 Aug 2019   Prob (F-statistic):               0.00
Time:                        12:02:19   Log-Likelihood:                -17398.
No. Observations:                1460   AIC:                         3.483e+04
Df Residuals:                    1445   BIC:                         3.491e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const           2.689e+04   1.58e+04      1.707      0.088   -4016.933    5.78e+04
overallqual     1.746e+04   1210.071     14.433      0.000    1.51e+04    1.98e+04
grlivarea         47.3229      2.328     20.325      0.000      42.756      51.890
garagecars      1.159e+04   2857.073      4.058      0.000    5989.270    1.72e+04
garagearea        22.4719      9.710      2.314      0.021       3.424      41.519
kitchenqual_Fa -5.033e+04   8337.143     -6.037      0.000   -6.67e+04    -3.4e+04
kitchenqual_Gd -4.014e+04   4790.348     -8.379      0.000   -4.95e+04   -3.07e+04
kitchenqual_TA -4.955e+04   5368.686     -9.229      0.000   -6.01e+04    -3.9e+04
exterqual_Fa   -4.701e+04   1.31e+04     -3.590      0.000   -7.27e+04   -2.13e+04
exterqual_Gd   -4.128e+04   6394.290     -6.456      0.000   -5.38e+04   -2.87e+04
exterqual_TA   -5.397e+04   7061.179     -7.644      0.000   -6.78e+04   -4.01e+04
mszoning_FV     2.786e+04    1.3e+04      2.145      0.032    2385.616    5.33e+04
mszoning_RH     2.196e+04   1.49e+04      1.476      0.140   -7216.155    5.11e+04
mszoning_RL     3.928e+04   1.21e+04      3.256      0.001    1.56e+04    6.29e+04
mszoning_RM     1.584e+04   1.22e+04      1.298      0.195   -8106.710    3.98e+04
==============================================================================
Omnibus:                      392.225   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            31451.825
Skew:                           0.027   Prob(JB):                         0.00
Kurtosis:                      25.738   Cond. No.                     4.84e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.84e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Goodness of Fit metrics
R-squared = 0.792 <br>
Adjusted R-squared = 0.790 <br>
F-test = 567.3, p-value = 0.00 <br>
AIC = 34830 <br>
BIC = 34910 <br><br>

#### Model Comparison
Compared to the first model, this model performs better. The F-statistic p-value equals 0, the Adjusted R-squared value is higher, and the AIC/BIC scores are lower.  This model predicts 1.5% more of the explained variance in the target than the first model.  

In [7]:
#Setting additional variables for model 3: interaction term between overall qual and first floor square feet
housing_df['int_overallqual_firstflrsf'] = housing_df.overallqual * housing_df.firstflrsf
X = housing_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'int_overallqual_firstflrsf', 'firstflrsf'] + dummy_column_names]
Y = housing_df.saleprice

In [8]:
#Running regression model 3

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.797
Method:                 Least Squares   F-statistic:                     359.6
Date:                Sun, 18 Aug 2019   Prob (F-statistic):               0.00
Time:                        12:02:19   Log-Likelihood:                -17371.
No. Observations:                1460   AIC:                         3.478e+04
Df Residuals:                    1443   BIC:                         3.487e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================================
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
const                       5.072e+04   1.77e+04      2.860      0.004    1.59e+04    8.55e+04
overallqual                 9030.1714   2214.923      4.077      0.000    4685.358    1.34e+04
grlivarea                     42.2828      2.444     17.304      0.000      37.490      47.076
garagecars                  1.468e+04   2846.780      5.158      0.000    9098.570    2.03e+04
garagearea                     6.0743      9.789      0.621      0.535     -13.128      25.277
int_overallqual_firstflrsf     6.8728      1.606      4.281      0.000       3.723      10.022
firstflrsf                   -26.5219     11.350     -2.337      0.020     -48.787      -4.257
kitchenqual_Fa              -4.66e+04   8203.556     -5.680      0.000   -6.27e+04   -3.05e+04
kitchenqual_Gd             -3.425e+04   4792.268     -7.147      0.000   -4.36e+04   -2.48e+04
kitchenqual_TA              -4.42e+04   5334.945     -8.286      0.000   -5.47e+04   -3.37e+04
exterqual_Fa               -3.771e+04   1.29e+04     -2.917      0.004   -6.31e+04   -1.23e+04
exterqual_Gd               -2.965e+04   6567.692     -4.514      0.000   -4.25e+04   -1.68e+04
exterqual_TA               -4.113e+04   7264.623     -5.661      0.000   -5.54e+04   -2.69e+04
mszoning_FV                 3.549e+04   1.28e+04      2.767      0.006    1.03e+04    6.06e+04
mszoning_RH                 2.457e+04   1.46e+04      1.677      0.094   -4163.719    5.33e+04
mszoning_RL                 3.947e+04   1.19e+04      3.315      0.001    1.61e+04    6.28e+04
mszoning_RM                 1.772e+04    1.2e+04      1.476      0.140   -5826.777    4.13e+04
==============================================================================
Omnibus:                      752.800   Durbin-Watson:                   2.001
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            97418.023
Skew:                          -1.382   Prob(JB):                         0.00
Kurtosis:                      42.922   Cond. No.                     2.47e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.47e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Goodness of Fit metrics
R-squared = 0.800 <br>
Adjusted R-squared = 0.797 <br>
F-test = 567.3, p-value = 0.00 <br>
AIC = 34780 <br>
BIC = 34870 <br><br>

#### Model Comparison
Compared to the previous two models, this model performs the best. The F-statistic p-value equals 0, the Adjusted R-squared value is higher than both Models 1 and 2, and the AIC/BIC scores are lower.  